rust/src/index/vector/ivf.rs

```diff
@@ -325,16 +325,17 @@ impl Ivf {
         for i in 0..data.num_rows() {
             let vector = data.row(i).unwrap();
             let part_id = argmin(
-                dist_func(vector.values(), centroids.data().values(), dim)
+                dist_func(vector, centroids.data().values(), dim)
                     .unwrap()
                     .as_ref(),
             )
             .unwrap();
             part_id_builder.append_value(part_id);
             let cent = centroids.row(part_id as usize).unwrap();
-            let residual = subtract_dyn(&vector, &cent)?;
-            let resi_arr: &Float32Array = as_primitive_array(&residual);
-            residual_builder.append_slice(resi_arr.values());
+            unsafe {
+                residual_builder
+                    .append_trusted_len_iter(vector.iter().zip(cent.iter()).map(|(v, c)| v - c))
```

Comment:
If somehow vector and centroid are incompatible does this segfault or just panic or does it fail silently?

Reply:
I can check the length of the two slices.  Their types should be the same to be compiled tho.

Reply:
I feel that this should be `assert` or `debug_assert` tho. no recoverable bug.